paho/client.go

```diff
@@ -0,0 +1,375 @@
+package paho
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	p "github.com/eclipse/paho.mqtt.golang/packets"
+)
+
+// Client is the struct representing an MQTT client
+type Client struct {
+	sync.Mutex
+	Stop          chan struct{}
+	Workers       sync.WaitGroup
+	Conn          net.Conn
+	MIDs          MIDService
+	PingHandler   Pinger
+	Router        Router
+	Persistence   Persistence
+	PacketTimeout time.Duration
+	Disconnected  func(p.Disconnect)
+}
+
+// NewClient is used to create a new instance of an MQTT client. It
+// returns a pointer to the new client instance and an error.
+// opts is a variadic of functions that take a pointer to a Client and
+// returns an error. These functions are used to modify the properties
+// of the client (such as setting the PingHandler, Router, etc), a
+// selection of such functions are provided in this library.
+func NewClient(opts ...func(*Client) error) (*Client, error) {
+	debug.Println("Creating new client")
+	c := &Client{
+		Stop: make(chan struct{}),
+	}
+
+	debug.Println("Setting options")
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+
+	if c.PingHandler == nil {
+		c.PingHandler = &PingHandler{
+			pingFailHandler: func(e error) {
+				c.Error(e)
+			},
+		}
+	}
+
+	if c.Persistence == nil {
+		c.Persistence = &noopPersistence{}
+	}
+
+	if c.MIDs == nil {
+		c.MIDs = &MIDs{index: make(map[uint16]*CPContext)}
+	}
+
+	if c.PacketTimeout == 0 {
+		c.PacketTimeout = 10 * time.Second
+	}
+
+	return c, nil
+}
+
+// Connect is used to connect the client to a server. It presumes that
+// the Client instance already has a working network connection.
+// The function takes a pre-prepared Connect packet, and uses that to
+// establish an MQTT connection. Assuming the connection completes
+// successfully the rest of the client is initiated and the Connack
+// returned. Otherwise the failure Connack (if there is one) is returned
+// along with an error indicating the reason for the failure to connect.
+func (c *Client) Connect(cp *p.Connect) (*p.Connack, error) {
```

Comment:
Why not pass directly the connect packet options directly to the Connect() func. Has the connect packet another use than to be passed to Connect()? Otherwise, I feel it's an implementation detail that burdens the API.

Same comment for Disconnect(), Publish(), Subscribe()

Reply:
For flexibility, the option funcs to configure the packets don't currently cover everything you might want to do, and as the fields of the packets are all exported from the `packets` library you are just as free to edit them directly when constructing a packet.

Reply:
And why not directly parameters to the function?
What about zero values with sensible default ?